sql: replace sampledPlanKey with statement fingerprint ID in ssmemstorage

gtr opened the pull request:
Closes [#69036](https://github.com/cockroachdb/cockroach/issues/69036).

Previously, the `stmtKey` struct contained the `sampledPlanKey` struct
which was redundant because a statement fingerprint ID has a one-to-one
mapping to a `sampledPlanKey`. This change replaces the `sampledPlanKey`
struct with a statement fingerprint ID and refactors the code in
`ssmemstorage` accordingly.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77154)
<!-- Reviewable:end -->

Azhng commented:
Thanks for taking up this! The code has changed quite a bit since the original issue was filed. I posted an update there.

Azhng reviewed (commented):
Nice refactoring. Make sure the CI is green and tests are passing. Also make sure squash your commits since reviewable can handle force pushes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-:-Mx6YPu845uMCFgRdBJc:b-1sgwbb)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @gtr and @maryliag)

___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6YvO01w-YOTTn48lh:-Mx6YvO01w-YOTTn48li:b-i1yp55) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L50)):*
> ```Go
> 
> func (s stmtKey) String() string {
> 	return strconv.FormatUint(uint64(s.statementFingerprintID), 10)
> ```

Normally we use `hex` [encoding](https://github.com/cockroachdb/cockroach/blob/08b663be38544a25b679bd077d6aca0d4e7fb8d9/pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil/json_encoding.go#L256) when we want to stringify fingerprint ID into a user readable string.
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6Wu4g3iPsp9IDJCjo:-Mx6Wu4g3iPsp9IDJCjp:b-1lvz32) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L245)):*
> ```Go
> 			statistics[i].Key.KeyData.ImplicitTxn,
> 			statistics[i].Key.KeyData.Database,
> 		)
> ```

Hmm I think the statement fingerprint ID is already in the payload here. You can get it through `statistics[i].ID` i think.
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 650 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6XlBy2xt3solyaIH6:-Mx6XlBy2xt3solyaIH7:b-skyav8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L650)):*
> ```Go
> 				statistics.Key.ImplicitTxn,
> 				statistics.Key.Database,
> 			)
> ```

`CollectedStatementStatistics` should also contain the field for fingerprint ID, so this construction is not necessary. 
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go, line 190 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6Y6vXArpAp7FXTi3d:-Mx6Y6vXArpAp7FXTi3e:bfv9kj9) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go#L190)):*
> ```Go
> 	fingerprint string, implicitTxn bool, database string,
> ) bool {
> 	lastSampled := s.getLogicalPlanLastSampled(roachpb.ConstructStatementFingerprintID(fingerprint, false, implicitTxn, database))
> ```

nit: wrap the long lines here. Also add an inline comment for the `false` literal
___
*[pkg/sql/sqlstats/ssmemstorage/utils.go, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6YFrR9p6iScVerfN3:-Mx6YFrR9p6iScVerfN4:bm9vst1) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/utils.go#L26)):*
> ```Go
> }
> func (s stmtList) Less(i, j int) bool {
> 	return s[i].statementFingerprintID < s[j].statementFingerprintID
> ```

Probably not a big deal, but note that the original compare function also took `transactionFingerprintID` into account. 


<!-- Sent from Reviewable.io -->

Commit:
sql: replace sampledPlanKey with statement fingerprint ID in ssmemstorage

Previously, the `stmtKey` struct contained the `sampledPlanKey` struct
which was redundant because a statement fingerprint ID has a one-to-one
mapping to a `sampledPlanKey`. This change replaces the `sampledPlanKey`
struct with a statement fingerprint ID and refactors the code in
`ssmemstorage` accordingly.

Release note: None

gtr reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77154)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng, @gtr, and @maryliag)

___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 50 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6YvO01w-YOTTn48lh:-MxzXhK66Gt3sppu9jBE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L50)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Normally we use `hex` [encoding](https://github.com/cockroachdb/cockroach/blob/08b663be38544a25b679bd077d6aca0d4e7fb8d9/pkg/sql/sqlstats/persistedsqlstats/sqlstatsutil/json_encoding.go#L256) when we want to stringify fingerprint ID into a user readable string.
</blockquote></details>

Done.
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6Wu4g3iPsp9IDJCjo:-MxzXiXm8h9nPwwunCvc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L245)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Hmm I think the statement fingerprint ID is already in the payload here. You can get it through `statistics[i].ID` i think.
</blockquote></details>

Done.
___
*[pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go, line 650 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6XlBy2xt3solyaIH6:-MxzXj_fEs-ZbVrWqfkF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/ss_mem_storage.go#L650)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

`CollectedStatementStatistics` should also contain the field for fingerprint ID, so this construction is not necessary. 
</blockquote></details>

Done.
___
*[pkg/sql/sqlstats/ssmemstorage/utils.go, line 26 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/77154#-Mx6YFrR9p6iScVerfN3:-MxzXlPWD_vEztHrAxqB:bmzqh09) ([raw file](https://github.com/cockroachdb/cockroach/blob/b69a36bb17443d0aa9f2efb59866b46e58e84e6d/pkg/sql/sqlstats/ssmemstorage/utils.go#L26)):*
<details><summary><i>Previously, Azhng (Archer Zhang) wrote…</i></summary><blockquote>

Probably not a big deal, but note that the original compare function also took `transactionFingerprintID` into account. 
</blockquote></details>

Left `transactionFingerprintID` as a tie-breaker thanks :).


<!-- Sent from Reviewable.io -->
